mm2src/coins/nft/storage/sql_storage.rs

```diff
@@ -165,47 +209,179 @@ fn nft_history_table_builder_preimage(
+        }
+        if filters.exclude_phishing {
+            sql_builder.and_where("possible_phishing == 0");
+        }
     }
     drop_mutability!(sql_builder);
     Ok(sql_builder)
 }
 
-fn finalize_nft_list_sql_builder(
-    mut sql_builder: SqlBuilder,
-    offset: usize,
-    limit: usize,
-) -> MmResult<String, SqlError> {
+fn finalize_sql_builder(mut sql_builder: SqlBuilder, offset: usize, limit: usize) -> MmResult<String, SqlError> {
     let sql = sql_builder
-        .field("nft_list.details_json")
+        .field("*")
         .offset(offset)
         .limit(limit)
         .sql()
         .map_err(|e| SqlError::ToSqlConversionFailure(e.into()))?;
     Ok(sql)
 }
 
-fn finalize_nft_history_sql_builder(
-    mut sql_builder: SqlBuilder,
-    offset: usize,
-    limit: usize,
-) -> MmResult<String, SqlError> {
-    let sql = sql_builder
-        .field("nft_history.details_json")
-        .offset(offset)
-        .limit(limit)
-        .sql()
-        .map_err(|e| SqlError::ToSqlConversionFailure(e.into()))?;
-    Ok(sql)
+fn get_and_parse<T: FromStr>(row: &Row<'_>, column: &str) -> Result<T, SqlError> {
+    let value_str: String = row.get(column)?;
+    value_str.parse().map_err(|_| SqlError::from(FromSqlError::InvalidType))
 }
 
 fn nft_from_row(row: &Row<'_>) -> Result<Nft, SqlError> {
-    let json_string: String = row.get(0)?;
-    json::from_str(&json_string).map_err(|e| SqlError::FromSqlConversionFailure(0, Type::Text, Box::new(e)))
+    let token_address = get_and_parse(row, "token_address")?;
+    let token_id = get_and_parse(row, "token_id")?;
+    let chain = get_and_parse(row, "chain")?;
+    let amount = get_and_parse(row, "amount")?;
+    let block_number: u64 = row.get("block_number")?;
+    let contract_type = get_and_parse(row, "contract_type")?;
+    let possible_spam: i32 = row.get("possible_spam")?;
+    let possible_phishing: i32 = row.get("possible_phishing")?;
+    let collection_name: Option<String> = row.get("collection_name")?;
+    let symbol: Option<String> = row.get("symbol")?;
+    let token_uri: Option<String> = row.get("token_uri")?;
+    let token_domain: Option<String> = row.get("token_domain")?;
+    let metadata: Option<String> = row.get("metadata")?;
+    let last_token_uri_sync: Option<String> = row.get("last_token_uri_sync")?;
+    let last_metadata_sync: Option<String> = row.get("last_metadata_sync")?;
+    let raw_image_url: Option<String> = row.get("raw_image_url")?;
+    let image_url: Option<String> = row.get("image_url")?;
+    let image_domain: Option<String> = row.get("image_domain")?;
+    let token_name: Option<String> = row.get("token_name")?;
+    let description: Option<String> = row.get("description")?;
+    let attributes_str: Option<String> = row.get("attributes")?;
+    let attributes: Option<Json> = attributes_str
+        .as_deref()
+        .map(json::from_str)
+        .transpose()
+        .map_err(|e| SqlError::FromSqlConversionFailure(21, Type::Text, Box::new(e)))?;
```

Comment:
Is 21 the column id or sql error code?

Reply:
usize should be number of particular column requested

Reply:
I guess you mean colum id which also used for prepared statements. I wasn't sure what 21 was.

Reply:
yeah, I also think its column id. bcz error doc comment says `Error when the value of a particular column is requested, but it cannot be converted to the requested Rust type.`